### PR TITLE
BUILD-281: wire in OCM feature gate config fields to allow eventual use of CSI volumes when BUILD-275 lands

### PR DESCRIPTION
**Pull request by gabemontero**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -15,53 +15,57 @@ require (
...
+	k8s.io/controller-manager v0.22.1
 	k8s.io/klog/v2 v2.9.0
-	k8s.io/kubectl v0.22.0-rc.0
-	k8s.io/kubernetes v1.22.0
+	k8s.io/kubectl v0.22.1
+	k8s.io/kubernetes v1.22.1
 	k8s.io/utils v0.0.0-20210707171843-4b05e18ac7d9
 )
 
+replace github.com/openshift/api => github.com/gabemontero/api v0.0.0-20210910154316-eaa29e1f16df
```
Comment (gabemontero) [inline go.mod:40]: this captures the latest state of openshift/api#1007

when that merges, we'll just change the version for openshift/api up above

File: go.mod [hunk 2]
```diff
@@ -15,53 +15,57 @@ require (
...
+
+replace github.com/openshift/library-go => github.com/gabemontero/library-go v0.0.0-20210910154638-8565ba189275
```
Comment (gabemontero) [inline go.mod:42]: this captures the latest state of openshift/library-go#1201

when that merges, we'll just change the version for openshift/library-go up above

File: pkg/cmd/controller/build.go [hunk 3]
```diff
@@ -49,6 +50,23 @@ func RunBuildController(ctx *ControllerContext) (bool, error) {
...
+	fg := ctx.OpenshiftControllerConfig.FeatureGates
+	// 		featuregates.NewObserveFeatureFlagsFunc(sets.NewString("BuildCSIVolumes"), sets.String{}, []string{"featureGates"}),
+	foundBuildCSIVolumes := false
+	foundCSIDriverSharedResource := false
+	if fg != nil && len(fg) > 0 {
+		for _, a := range fg {
+			for _, v := range a {
+				switch strings.TrimSpace(v) {
+				case "CSIDriverSharedResource":
+					foundCSIDriverSharedResource = true
```
Comment (adambkaplan) [inline pkg/cmd/controller/build.go:62]: We don't need to check this feature gate - this gate is used to install the shared resource CSI driver on the cluster.

File: pkg/cmd/controller/build.go [hunk 4]
```diff
@@ -65,11 +83,13 @@ func RunBuildController(ctx *ControllerContext) (bool, error) {
...
-			Image: imageTemplate.ExpandOrDie("docker-builder"),
+			Image:                    imageTemplate.ExpandOrDie("docker-builder"),
+			BuildCSIVolumesAvailable: foundBuildCSIVolumes && foundCSIDriverSharedResource,
 		},
 		SourceBuildStrategy: &buildstrategy.SourceBuildStrategy{
-			Image:          imageTemplate.ExpandOrDie("docker-builder"),
-			SecurityClient: securityClient.SecurityV1(),
+			Image:                    imageTemplate.ExpandOrDie("docker-builder"),
+			SecurityClient:           securityClient.SecurityV1(),
+			BuildCSIVolumesAvailable: foundBuildCSIVolumes && foundCSIDriverSharedResource,
```
Comment (adambkaplan) [inline pkg/cmd/controller/build.go:92]: Only check the build flag:
```suggestion
			BuildCSIVolumesAvailable: foundBuildCSIVolumes,
```

File: pkg/cmd/controller/build.go (hunk 3 above)
Comment (gabemontero) [reply pkg/cmd/controller/build.go]: updated @adambkaplan PTAL

File: pkg/cmd/controller/build.go (hunk 4 above)
Comment (gabemontero) [reply pkg/cmd/controller/build.go]: updated @adambkaplan PTAL

**Comment by gabemontero** [comment]
unrelated fails

/retest

**Comment by gabemontero** [comment]
/assign @adambkaplan 

OK this has been updated with the merged versions of openshift/api and openshift/library-go and is ready for approve and/or review

**Comment by gabemontero** [comment]
unit test *appears* to be unrelated flake that passes for me locally ... let's see

/test unit

**Comment by gabemontero** [comment]
install time networking fail in e2e-aws-proxy

will wait on retest in case our aws is overloaded

**Comment by gabemontero** [comment]
inability to access registry.redhat.io with latest e2e-gcp-builds

/retest

File: pkg/build/controller/build/build_controller.go [hunk 5]
```diff
@@ -190,6 +190,7 @@ type BuildController struct {
 	buildDefaults            builddefaults.BuildDefaults
 	buildOverrides           buildoverrides.BuildOverrides
 	internalRegistryHostname string
+	buildCSIVolumesAvailable bool
```
Comment (adambkaplan) [inline pkg/build/controller/build/build_controller.go:193]:
```suggestion
	buildCSIVolumesEnabled bool
```

File: pkg/build/controller/strategy/docker.go [hunk 6]
```diff
@@ -26,7 +26,8 @@ func init() {
 
 // DockerBuildStrategy creates a Docker build using a Docker builder image.
 type DockerBuildStrategy struct {
-	Image string
+	Image                    string
+	BuildCSIVolumesAvailable bool
```
Comment (adambkaplan) [inline pkg/build/controller/strategy/docker.go:30]:
```suggestion
	BuildCSIVolumesEnabled bool
```

File: pkg/build/controller/strategy/docker.go [hunk 7]
```diff
@@ -214,7 +215,7 @@ func (bs *DockerBuildStrategy) CreateBuildPod(build *buildv1.Build, additionalCA
 	setupContainersStorage(pod, &pod.Spec.Containers[0]) // for unprivileged builds
 	// setupContainersNodeStorage(pod, &pod.Spec.Containers[0]) // for privileged builds
 	setupBlobCache(pod)
-	if err := setupBuildVolumes(pod, build.Spec.Strategy.DockerStrategy.Volumes); err != nil {
+	if err := setupBuildVolumes(pod, build.Spec.Strategy.DockerStrategy.Volumes, bs.BuildCSIVolumesAvailable); err != nil {
```
Comment (adambkaplan) [inline pkg/build/controller/strategy/docker.go:218]:
```suggestion
	if err := setupBuildVolumes(pod, build.Spec.Strategy.DockerStrategy.Volumes, bs.BuildCSIVolumesEnabled); err != nil {
```

File: pkg/build/controller/strategy/sti.go [hunk 8]
```diff
@@ -18,8 +18,9 @@ import (
 
 // SourceBuildStrategy creates STI(source to image) builds
 type SourceBuildStrategy struct {
-	Image          string
-	SecurityClient securityclient.SecurityV1Interface
+	Image                    string
+	SecurityClient           securityclient.SecurityV1Interface
+	BuildCSIVolumesAvailable bool
```
Comment (adambkaplan) [inline pkg/build/controller/strategy/sti.go:23]:
```suggestion
	BuildCSIVolumesEnabled bool
```

File: pkg/build/controller/strategy/sti.go [hunk 9]
```diff
@@ -221,7 +222,7 @@ func (bs *SourceBuildStrategy) CreateBuildPod(build *buildv1.Build, additionalCA
 	setupContainersStorage(pod, &pod.Spec.Containers[0]) // for unprivileged builds
 	// setupContainersNodeStorage(pod, &pod.Spec.Containers[0]) // for privileged builds
 	setupBlobCache(pod)
-	if err := setupBuildVolumes(pod, build.Spec.Strategy.SourceStrategy.Volumes); err != nil {
+	if err := setupBuildVolumes(pod, build.Spec.Strategy.SourceStrategy.Volumes, bs.BuildCSIVolumesAvailable); err != nil {
```
Comment (adambkaplan) [inline pkg/build/controller/strategy/sti.go:225]:
```suggestion
	if err := setupBuildVolumes(pod, build.Spec.Strategy.SourceStrategy.Volumes, bs.BuildCSIVolumesEnabled); err != nil {
```

File: pkg/build/controller/strategy/util.go [hunk 10]
```diff
@@ -668,7 +668,9 @@ func setupBlobCache(pod *corev1.Pod) {
 }
 
 // setupBuildVolumes sets up user defined BuildVolumes
-func setupBuildVolumes(pod *corev1.Pod, buildVolumes []buildv1.BuildVolume) error {
+func setupBuildVolumes(pod *corev1.Pod, buildVolumes []buildv1.BuildVolume, buildCSIVolumes bool) error {
```
Comment (adambkaplan) [inline pkg/build/controller/strategy/util.go:671]:
```suggestion
func setupBuildVolumes(pod *corev1.Pod, buildVolumes []buildv1.BuildVolume, csiVolumesEnabled bool) error {
```

File: pkg/cmd/controller/build.go [hunk 11]
```diff
@@ -50,6 +51,17 @@ func RunBuildController(ctx *ControllerContext) (bool, error) {
...
+	fg := ctx.OpenshiftControllerConfig.FeatureGates
+	foundBuildCSIVolumes := false
+	if fg != nil {
+		for _, v := range fg {
+			v = strings.TrimSpace(v)
+			if v == "BuildCSIVolumes=true" {
+				foundBuildCSIVolumes = true
+			}
+		}
+	}
```
Comment (adambkaplan) [inline pkg/cmd/controller/build.go:63]:
```suggestion
	csiVolumesEnabled := false
	if fg != nil {
		for _, v := range fg {
			v = strings.TrimSpace(v)
			if v == "BuildCSIVolumes=true" {
				csiVolumesEnabled = true
			}
		}
	}
```

File: pkg/cmd/controller/build.go [hunk 12]
```diff
@@ -67,11 +79,13 @@ func RunBuildController(ctx *ControllerContext) (bool, error) {
 		KubeClient:                         externalKubeClient,
 		BuildClient:                        buildClient,
 		DockerBuildStrategy: &buildstrategy.DockerBuildStrategy{
-			Image: imageTemplate.ExpandOrDie("docker-builder"),
+			Image:                    imageTemplate.ExpandOrDie("docker-builder"),
+			BuildCSIVolumesAvailable: foundBuildCSIVolumes,
```
Comment (adambkaplan) [inline pkg/cmd/controller/build.go:83]:
```suggestion
			BuildCSIVolumesEnabled: csiVolumesEnabled,
```

File: pkg/cmd/controller/build.go [hunk 13]
```diff
@@ -67,11 +79,13 @@ func RunBuildController(ctx *ControllerContext) (bool, error) {
...
 		},
 		SourceBuildStrategy: &buildstrategy.SourceBuildStrategy{
-			Image:          imageTemplate.ExpandOrDie("docker-builder"),
-			SecurityClient: securityClient.SecurityV1(),
+			Image:                    imageTemplate.ExpandOrDie("docker-builder"),
+			SecurityClient:           securityClient.SecurityV1(),
+			BuildCSIVolumesAvailable: foundBuildCSIVolumes,
```
Comment (adambkaplan) [inline pkg/cmd/controller/build.go:88]:
```suggestion
			BuildCSIVolumesEnabled: csiVolumesEnabled,
```

**Review by adambkaplan — COMMENTED** [review]
/approve

Generally looks good. Main suggestion is to use the phrase "CSIVolumesEnabled" instead of "Available" - I think this is better communicates intent.

**Comment by gabemontero** [comment]
renames pulled and re pushed, commits squashed @adambkaplan 

ptal / thanks

**Review by adambkaplan — APPROVED** [review]
/lgtm

**Pull request merged**